### ISSUE TITLE
[BUG] tap.before/tap.teardown are weird when import and requiring `tap`

### ISSUE THREAD

**Issue by derhuerst** [tapjs/tapjs#1005]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Have you read the `CONTRIBUTING` guide on posting bugs, and `CODE_OF_CONDUCT`?

- [X] yes I read the things

### This issue exists in the latest tap version

- [X] I am using the latest tap

### Description

It seems like `tap.before()` & `tap.after()`/`tap.teardown()` behave differently when I call them on the object returned by `require` when tap has been `import`ed before.

If `require` is used, omitting the `before()` also breaks `tap.after()`/`tap.teardown()`.

### Reproduction

```shell
mkdir tap-bug
cd tap-bug
npm init --yes
# put type: module in package.json
npm i tap
# put tap config in package.json
# put test.js
```

### `test.js`

```js
import tap from 'tap';
import {createRequire} from 'module';

{
	// const tap = createRequire(import.meta.url)('tap');

	tap.before(async () => {
		console.log('before');
	});
	tap.teardown(async () => {
		console.log('global teardown');
	});
}

tap.test('some test', async (t) => {
	await new Promise(r => setTimeout(r, 300));
	t.teardown(async () => {
		console.log('test-specific teardown');
	});
	t.end();
});
```

### without `require()`, with `before()`

```
before
TAP version 14
# Subtest: some test
    1..0
test-specific teardown
ok 1 - some test # time=306.614ms

1..1
global teardown
# { total: 1, pass: 1 }
# time=314.954ms
```

### without `require()`, without `before()`

```
TAP version 14
# Subtest: some test
    1..0
test-specific teardown
ok 1 - some test # time=308.365ms

1..1
global teardown
# { total: 1, pass: 1 }
# time=316.118ms
```

### with `require()`, with `before()`

Note that
- `before` appears before `TAP version 14`
- `TAP version 14` appears twice
- `no tests found` appears
- `global teardown` appears before the actual test.

```
before
TAP version 14
TAP version 14
1..0 # no tests found
global teardown
# { total: 0, pass: 0 }
# time=6.119ms
# Subtest: some test
    1..0
test-specific teardown
ok 1 - some test # time=308.49ms

1..1
# { total: 1, pass: 1 }
# time=345.437ms
```

### with `require()`, without `before()`

Note that
- `TAP version 14` appears only once
- `global teardown` missing!

```
TAP version 14
# Subtest: some test
    1..0
test-specific teardown
ok 1 - some test # time=308.937ms

1..1
# { total: 1, pass: 1 }
# time=344.677ms
```

### Environment

```shell
which tap
# ./node_modules/.bin/tap
npm ls tap
# tap-bug@1.0.0 /Users/j/playground/tap-bug
└── tap@18.7.0
tap versions
# tap: 18.7.0
# "@tapjs/config": 2.4.15
# "@tapjs/core": 1.5.0
# "@tapjs/run": 1.5.0
# "@tapjs/stack": 1.2.7
# "@tapjs/test": 1.4.0
# tap-parser: 15.3.1
# tap-yaml: 2.2.1
# tcompare: 6.4.5
# plugins:
#   "@tapjs/after": 1.1.18
#   "@tapjs/after-each": 1.1.18
#   "@tapjs/asserts": 1.1.18
#   "@tapjs/before": 1.1.18
#   "@tapjs/before-each": 1.1.18
#   "@tapjs/filter": 1.2.18
#   "@tapjs/fixture": 1.2.18
#   "@tapjs/intercept": 1.2.18
#   "@tapjs/mock": 1.3.0
#   "@tapjs/node-serialize": 1.3.0
#   "@tapjs/snapshot": 1.2.18
#   "@tapjs/spawn": 1.1.18
#   "@tapjs/stdin": 1.1.18
#   "@tapjs/typescript": 1.4.0
#   "@tapjs/worker": 1.1.18
tap config list
# # vim: set filetype=yaml :
# 
# # from package.json
# allow-empty-coverage: true
# disable-coverage: true
# plugin:
#   - "!@tapjs/typescript"
#   - "!@tapjs/mock"
#   - "!@tapjs/snapshot"
#   - "!@tapjs/intercept"
#   - "!@tapjs/fixture"
# 
# # env, cli, and defaults
# color: false
# coverage-report:
#   - text
# exclude:
#   - "**/@(fixture*(s)|dist)/**"
# include:
#   - "**/@(test?(s)|__test?(s)__)/**/*.@(js|cjs|mjs|tap|cts|jsx|mts|ts|tsx)"
#   - "**/*.@(test?(s)|spec).@(js|cjs|mjs|tap|cts|jsx|mts|ts|tsx)"
#   - "**/test?(s).@(js|cjs|mjs|tap|cts|jsx|mts|ts|tsx)"
# jobs: 6
# reporter: tap
# snapshot-clean-cwd: true
# timeout: 30
tap plugin list
# @tapjs/after
# @tapjs/after-each
# @tapjs/asserts
# @tapjs/before
# @tapjs/before-each
# @tapjs/filter
# @tapjs/node-serialize
# @tapjs/spawn
# @tapjs/stdin
# @tapjs/worker
uname -a
# Darwin jannis-mba2.local 22.6.0 Darwin Kernel Version 22.6.0: Sun Dec 17 22:14:25 PST 2023; root:xnu-8796.141.3.703.2~2/RELEASE_ARM64_T8112 arm64
```

### `npm ls --depth 2`

```
tap-bug@1.0.0 /Users/j/playground/tap-bug
└─┬ tap@18.7.0
  ├─┬ @tapjs/after-each@1.1.18
  │ ├── @tapjs/core@1.5.0 deduped
  │ └── function-loop@4.0.0
  ├─┬ @tapjs/after@1.1.18
  │ ├── @tapjs/core@1.5.0 deduped
  │ └── is-actual-promise@1.0.1
  ├─┬ @tapjs/asserts@1.1.18
  │ ├── @tapjs/core@1.5.0 deduped
  │ ├── @tapjs/stack@1.2.7
  │ ├── is-actual-promise@1.0.1 deduped
  │ ├── tcompare@6.4.5
  │ └── trivial-deferred@2.0.0
  ├─┬ @tapjs/before-each@1.1.18
  │ ├── @tapjs/core@1.5.0 deduped
  │ └── function-loop@4.0.0 deduped
  ├─┬ @tapjs/before@1.1.18
  │ ├── @tapjs/core@1.5.0 deduped
  │ └── is-actual-promise@1.0.1 deduped
  ├─┬ @tapjs/core@1.5.0
  │ ├── @tapjs/processinfo@3.1.6
  │ ├── @tapjs/stack@1.2.7 deduped
  │ ├── @tapjs/test@1.4.0 deduped
  │ ├── async-hook-domain@4.0.1
  │ ├── diff@5.2.0
  │ ├── is-actual-promise@1.0.1 deduped
  │ ├── minipass@7.0.4
  │ ├── signal-exit@4.1.0
  │ ├── tap-parser@15.3.1
  │ ├── tap-yaml@2.2.1
  │ ├── tcompare@6.4.5 deduped
  │ └── trivial-deferred@2.0.0 deduped
  ├─┬ @tapjs/filter@1.2.18
  │ └── @tapjs/core@1.5.0 deduped
  ├─┬ @tapjs/fixture@1.2.18
  │ ├── @tapjs/core@1.5.0 deduped
  │ ├── mkdirp@3.0.1
  │ └── rimraf@5.0.5
  ├─┬ @tapjs/intercept@1.2.18
  │ ├── @tapjs/after@1.1.18 deduped
  │ ├── @tapjs/core@1.5.0 deduped
  │ └── @tapjs/stack@1.2.7 deduped
  ├─┬ @tapjs/mock@1.3.0
  │ ├── @tapjs/after@1.1.18 deduped
  │ ├── @tapjs/core@1.5.0 deduped
  │ ├── @tapjs/stack@1.2.7 deduped
  │ ├── resolve-import@1.4.5 deduped
  │ └── walk-up-path@3.0.1
  ├─┬ @tapjs/node-serialize@1.3.0
  │ ├── @tapjs/core@1.5.0 deduped
  │ ├── @tapjs/error-serdes@1.2.1
  │ ├── @tapjs/stack@1.2.7 deduped
  │ └── tap-parser@15.3.1 deduped
  ├─┬ @tapjs/run@1.5.0
  │ ├── @tapjs/after@1.1.18 deduped
  │ ├── @tapjs/before@1.1.18 deduped
  │ ├── @tapjs/config@2.4.15
  │ ├── @tapjs/core@1.5.0 deduped
  │ ├── @tapjs/processinfo@3.1.6 deduped
  │ ├── @tapjs/reporter@1.3.16
  │ ├── @tapjs/spawn@1.1.18 deduped
  │ ├── @tapjs/stdin@1.1.18 deduped
  │ ├── @tapjs/test@1.4.0 deduped
  │ ├── c8@8.0.1
  │ ├── chalk@5.3.0
  │ ├── chokidar@3.6.0
  │ ├── foreground-child@3.1.1
  │ ├── glob@10.3.10
  │ ├── minipass@7.0.4 deduped
  │ ├── mkdirp@3.0.1 deduped
  │ ├── opener@1.5.2
  │ ├── pacote@17.0.6
  │ ├── resolve-import@1.4.5 deduped
  │ ├── rimraf@5.0.5 deduped
  │ ├── semver@7.6.0
  │ ├── signal-exit@4.1.0 deduped
  │ ├── tap-parser@15.3.1 deduped
  │ ├── tap-yaml@2.2.1 deduped
  │ ├── tcompare@6.4.5 deduped
  │ ├── trivial-deferred@2.0.0 deduped
  │ └── which@4.0.0
  ├─┬ @tapjs/snapshot@1.2.18
  │ ├── @tapjs/core@1.5.0 deduped
  │ ├── is-actual-promise@1.0.1 deduped
  │ ├── tcompare@6.4.5 deduped
  │ └── trivial-deferred@2.0.0 deduped
  ├─┬ @tapjs/spawn@1.1.18
  │ └── @tapjs/core@1.5.0 deduped
  ├─┬ @tapjs/stdin@1.1.18
  │ └── @tapjs/core@1.5.0 deduped
  ├─┬ @tapjs/test@1.4.0
  │ ├── @isaacs/ts-node-temp-fork-for-pr-2009@10.9.5
  │ ├── @tapjs/after-each@1.1.18 deduped
  │ ├── @tapjs/after@1.1.18 deduped
  │ ├── @tapjs/asserts@1.1.18 deduped
  │ ├── @tapjs/before-each@1.1.18 deduped
  │ ├── @tapjs/before@1.1.18 deduped
  │ ├── @tapjs/core@1.5.0 deduped
  │ ├── @tapjs/filter@1.2.18 deduped
  │ ├── @tapjs/fixture@1.2.18 deduped
  │ ├── @tapjs/intercept@1.2.18 deduped
  │ ├── @tapjs/mock@1.3.0 deduped
  │ ├── @tapjs/node-serialize@1.3.0 deduped
  │ ├── @tapjs/snapshot@1.2.18 deduped
  │ ├── @tapjs/spawn@1.1.18 deduped
  │ ├── @tapjs/stdin@1.1.18 deduped
  │ ├── @tapjs/typescript@1.4.0 deduped
  │ ├── @tapjs/worker@1.1.18 deduped
  │ ├── glob@10.3.10 deduped
  │ ├── jackspeak@2.3.6
  │ ├── mkdirp@3.0.1 deduped
  │ ├── resolve-import@1.4.5 deduped
  │ ├── rimraf@5.0.5 deduped
  │ ├── sync-content@1.0.2
  │ ├── tap-parser@15.3.1 deduped
  │ ├── tshy@1.11.1
  │ └── typescript@5.2.2
  ├─┬ @tapjs/typescript@1.4.0
  │ ├── @isaacs/ts-node-temp-fork-for-pr-2009@10.9.5 deduped
  │ └── @tapjs/core@1.5.0 deduped
  ├─┬ @tapjs/worker@1.1.18
  │ └── @tapjs/core@1.5.0 deduped
  └─┬ resolve-import@1.4.5
    ├── glob@10.3.10 deduped
    └── walk-up-path@3.0.1 deduped
```

**Comment by isaacs** [comment]
I believe you're running into dual module hazard. The tap loaded with `require` and the tap loaded with `import` are effectively two different modules. I can look into what might be done, but I think in the short term, the best bet is to use it with either import or require, but not both in the same test file.
